### PR TITLE
unify enroll and encrypt, add pin support

### PR DESCRIPTION
**Pull request by lnussel**

### CODE REVIEW

File: jeos-firstboot-enroll [hunk 1]
```diff
@@ -29,7 +32,14 @@ detect_luks2()
 
 enroll_systemd_firstboot() {
 	[ -e /usr/bin/systemd-cryptenroll ] || return 0
-	detect_luks2 || return 0
+	crypt_keyid="$(keyctl id %user:cryptenroll)"
+	[ -n "$crypt_keyid" ] || return 0
+
+	if ! detect_luks2; then
+		d --msgbox $"Error: recovery set but no encrypted disks found" 0 0
```
Comment (aplanas) [inline jeos-firstboot-enroll:42]: I would revert the condition, first check if there is a luk2 device and silently exit, and later check for the crypt_keyid, and if it is not present show this error. It is more clear.

File: jeos-firstboot-enroll [hunk 2]
```diff
@@ -29,7 +32,14 @@ detect_luks2()
...
+		unset crypt_keyid
```
Comment (aplanas) [inline jeos-firstboot-enroll:43]: in that way you can also avoid this

File: jeos-firstboot-enroll [hunk 3]
```diff
@@ -52,10 +62,32 @@ enroll_systemd_firstboot() {
...
+
+	if [ -n "$with_fido2" ] || [ -n "$with_tpm2" ]; then
+		d_styled --yesno $"Add additional encryption password?" 0 0 || return 0
+	fi
+
+	msg=$"Use root password as encryption password?"
+	if [ -n "$password" ] && d_styled --yesno "$msg" 0 0; then
+		crypt_pw="$password"
+	else
+		while true; do
```
Comment (aplanas) [inline jeos-firstboot-enroll:95]: extract this into a function?

File: jeos-firstboot-enroll [hunk 4]
```diff
@@ -52,10 +62,32 @@ enroll_systemd_firstboot() {
...
 	elif [ "$has_fido2" ]; then
-		dialog $dialog_alternate_screen --backtitle "$PRETTY_NAME" --yesno $"Unlock encrypted disk via FIDO2 token?" 0 0 && with_fido2=1
+		d_styled --yesno $"Unlock encrypted disk via FIDO2 token?" 0 0 && with_fido2=1
 	elif [ -n "$has_tpm2" ]; then
-		dialog $dialog_alternate_screen --backtitle "$PRETTY_NAME" --yesno $"Unlock encrypted disk via TPM?" 0 0 && with_tpm2="$has_tpm2"
+		d_styled --yesno $"Unlock encrypted disk via TPM?" 0 0 && with_tpm2="$has_tpm2"
+	fi
+
+	if [ -n "$with_fido2" ] || [ -n "$with_tpm2" ]; then
+		d_styled --yesno $"Add additional encryption password?" 0 0 || return 0
```
Comment (aplanas) [inline jeos-firstboot-enroll:88]: If the recovery password is there, why to ask to a lower entropy, less secure password?

There is already a change in cryptenroll that allows the user to use the TPM2 to add/remove a keyslots (like already can be done with the FIDO2 key), but even before that they can use the recover password to add an extra one if it is required.

File: jeos-firstboot-enroll [hunk 5]
```diff
@@ -52,10 +62,32 @@ enroll_systemd_firstboot() {
...
+	fi
+
+	msg=$"Use root password as encryption password?"
+	if [ -n "$password" ] && d_styled --yesno "$msg" 0 0; then
+		crypt_pw="$password"
+	else
+		while true; do
+			d --insecure --passwordbox  $"Enter encryption password" 0 0
+			if [ -z "$result" ]; then
+				d --aspect 29 --msgbox $"No encryption password set. You can add more keys manually using cryptsetup." 0 0
```
Comment (aplanas) [inline jeos-firstboot-enroll:98]: I would recommend cryptenroll

File: jeos-firstboot-enroll [hunk 6]
```diff
@@ -155,21 +187,61 @@ is_pcr_oracle() {
...
+	run issue-generator
+	[ -n "$dry" ] || cat "$issuefile"
+}
+
+add_password() {
+	[ -n "$crypt_pw" ] || return 0
+	local dev
+	for dev in "${luks2_devices[@]}"; do
+		echo "adding password to $dev"
+		echo -n "$crypt_pw" | run cryptsetup luksAddKey --verbose --batch-mode --force-password --key-file <(keyctl pipe "$crypt_keyid") "$dev"
```
Comment (aplanas) [inline jeos-firstboot-enroll:234]: why not to use cryptsetup? you can use the keyring for the keyid one, and a envvar to set the new one

File: jeos-firstboot-enroll [hunk 7]
```diff
@@ -206,3 +278,19 @@ enroll_post() {
 		run sdbootutil add-all-kernels --no-reuse-initrd
 	fi
 }
+
+enroll_jeos_config() {
+	d --insecure --passwordbox  $"Enter encryption password" 0 0
```
Comment (aplanas) [inline jeos-firstboot-enroll:308]: Should ask for the recovery password?

File: jeos-firstboot-enroll [hunk 8]
```diff
@@ -206,3 +278,19 @@ enroll_post() {
...
+	[ -n "$result" ] || return 0
+	echo -n "$result" | keyctl padd user cryptenroll @u
+
+	if ! detect_luks2; then
```
Comment (aplanas) [inline jeos-firstboot-enroll:312]: again, I would reverse the check: if there is no luks device -- nothing to do and then ask for the recovery password

File: jeos-firstboot-enroll [hunk 9]
```diff
@@ -67,6 +70,20 @@ enroll_systemd_firstboot() {
 		d_styled --yesno $"Unlock encrypted disk via TPM?" 0 0 && with_tpm2="$has_tpm2"
 	fi
 
+	if [ -n "$with_tpm2" ] && d_styled --yesno $"With TPM PIN protection?" 0 0; then
+		while true; do
+			d --insecure --passwordbox  $"Enter new PIN (actually just passphrase)" 0 0
```
Comment (aplanas) [inline jeos-firstboot-enroll:75]: there is a lot of y/n password, what about that if the PIN is empty, no PIN is set? The text can be "Enter blank PIN to automatic unlock" or something like that.

File: jeos-firstboot-enroll [hunk 10]
```diff
@@ -231,11 +250,11 @@ enroll_post() {
 enroll_tpm_and_fido() {
 	# For now is a first step before moving into fde-tools
 	local fde_cfg='/etc/sysconfig/fde-tools'
-	if [ -e "$cfg" ]; then
-		. "$cfg"
+	if [ -e "$fde_cfg" ]; then
```
Comment (aplanas) [inline jeos-firstboot-enroll:253]: should be part of the other PR?

File: jeos-firstboot-enroll [hunk 11]
```diff
@@ -264,6 +283,12 @@ enroll_tpm_and_fido() {
 			enroll_fido2 "$dev"
 		done
 	elif [ -n "$with_tpm2" ]; then
+		if [ -n "$crypt_tpm_pin" ]; then
+			# XXX ./src/cryptenroll/cryptenroll-tpm2.c lacks accept cached
+			#echo -n "$crypt_tpm_pin" | run keyctl padd user tpm2-pin @u
+			export NEWPIN="$crypt_tpm_pin"
```
Comment (aplanas) [inline jeos-firstboot-enroll:289]: Do not export the secret:
https://github.com/openSUSE/disk-encryption-tool/pull/12/files#diff-b3ccc79ad196422dc257215e20dc1d7a01b08b3e4d2ba43850689f15804042f3R134

**Review by aplanas — CHANGES_REQUESTED** [review]

File: jeos-firstboot-enroll (hunk 4 above)
Comment (aplanas) [reply jeos-firstboot-enroll:88]: Maybe the password can only be asked iff a TPM2 and FIDO2 key are missing, as a last resort mechanism

File: jeos-firstboot-enroll (hunk 1 above)
Comment (lnussel) [reply jeos-firstboot-enroll:42]: My thinking was that checking for the keyid is the cheaper check to return early. The error is not something for users anyway. More like an assert().

File: jeos-firstboot-enroll (hunk 7 above)
Comment (lnussel) [reply jeos-firstboot-enroll:308]: can be any passphrase actually

File: jeos-firstboot-enroll (hunk 9 above)
Comment (lnussel) [reply jeos-firstboot-enroll:75]: good idea

File: jeos-firstboot-enroll (hunk 1 above)
Comment (aplanas) [reply jeos-firstboot-enroll:42]: But the error is reversed: if there is a luks2 and there is no password then something did goes wrong (pbly the user executed jeosfb manually or `touch /var/lib/YaST2/reconfig_system`). The error should be that the keyid is missing and that the enrollment can fail.

The current check is: if there is keyid and no luks2 then I assume that there is a recovery key, but technically this is not an error

File: jeos-firstboot-enroll (hunk 4 above)
Comment (aplanas) [reply jeos-firstboot-enroll:88]: another idea is to replace the high entropy recovery password with the one in the env var, so the logic would still be consistent: format the luks device with the initial password (automatic or user defined), register the (TPM/FIDO xor user defined password)

**Comment by lnussel** [comment]
used a different approach with less yes/no

File: jeos-firstboot-enroll (hunk 9 above)
Comment (lnussel) [reply jeos-firstboot-enroll]: is now a separate item in the menu

File: jeos-firstboot-enroll (hunk 10 above)
Comment (lnussel) [reply jeos-firstboot-enroll:263]: was a mistake when merging diskencrypt and enroll. folded into parent commit now

File: jeos-firstboot-enroll (hunk 6 above)
Comment (lnussel) [reply jeos-firstboot-enroll:244]: for a separate commit then

File: jeos-firstboot-enroll (hunk 8 above)
Comment (lnussel) [reply jeos-firstboot-enroll]: yes. the check works differently now

File: disk-encryption-tool-dracut [hunk 12]
```diff
@@ -3,6 +3,22 @@
 exec < /dev/console >/dev/console 2>&1
 type getarg > /dev/null 2>&1 || . /lib/dracut-lib.sh
 
+# get systemd credential
+# https://systemd.io/CREDENTIALS/
+get_credential()
+{
+	local var="${1:?}"
+	local name="${2:?}"
+	[ -n "$CREDENTIALS_DIRECTORY" ] || return 1
```
Comment (aplanas) [inline disk-encryption-tool-dracut:12]: Was this always launched as a systemd service? I though that is was registered as a traditional dracut module.

File: jeos-firstboot-enroll [hunk 13]
```diff
@@ -12,197 +18,316 @@ FDE_SEAL_PCR_LIST="0,2,4,7,9"
...
+			list+=('password' $'Enroll extra password')
+		fi
+		[ -n "$list" ] || break
+
+		list+=('done' $'Done')
+
+		d --no-tags --default-item "${list[0]}" --menu $"Disk Encryption" 0 0 "$(menuheight ${#list[@]})" "${list[@]}"
+		if [ "$result" = 'done' ]; then
+			if [ -z "$crypt_pw" ] && [ -z "$with_fido2" ] && [ -z "$with_tpm2" ] && [ -z "$is_jeos_config" ]; then
+				d_styled --yesno $"Neither password, TPM nor FIDO2 entrolled. Unlocking disk will only work with recovery key. Is this intended?" 0 0 || continue
```
Comment (aplanas) [inline jeos-firstboot-enroll:79]: TPM2?

File: jeos-firstboot-enroll [hunk 14]
```diff
@@ -12,197 +18,316 @@ FDE_SEAL_PCR_LIST="0,2,4,7,9"
...
+				d --msgbox $"PINs don't match. Try again" 0 0
+			done
+
+		elif [ "$result" = 'root' ]; then
+			crypt_pw="$password"
+		elif [ "$result" = 'password' ]; then
+			while true; do
+				d --insecure --passwordbox  $"Enter encryption password" 0 0
+				if [ -z "$result" ]; then
+					d --aspect 29 --msgbox $"No encryption password set. You can add more keys manually using cryptsetup." 0 0
```
Comment (aplanas) [inline jeos-firstboot-enroll:105]: was commented before, but I would not recommend to use cryptsetup to the user, as should be systemd-cryptsetup now

File: jeos-firstboot-enroll [hunk 15]
```diff
@@ -12,197 +18,316 @@ FDE_SEAL_PCR_LIST="0,2,4,7,9"
...
-	[ "$result" = 'FIDO2' ] && with_fido2=1
-	[ "$result" = 'TPM2' ] && with_tpm2="$has_tpm2"
-    elif [ "$has_fido2" ]; then
-	dialog $dialog_alternate_screen --backtitle "$PRETTY_NAME" --yesno $"Unlock encrypted disk via FIDO2 token?" 0 0 && with_fido2=1
-    elif [ -n "$has_tpm2" ]; then
-	dialog $dialog_alternate_screen --backtitle "$PRETTY_NAME" --yesno $"Unlock encrypted disk via TPM?" 0 0 && with_tpm2="$has_tpm2"
-    fi
-    return 0
+
+	while true; do
```
Comment (aplanas) [inline jeos-firstboot-enroll:57]: Here start the interactive part, that now is not trivial.

Ideally all this should be unattended (the typical MicroOS use case will be in servers, with multiple installations). For a PoC was OK always to require interaction, but we should move out from it now.

In the same way that rd.credentials was added for this set, can we add something like rd.interactive to enter this codepath? If is missing and the TPM2 is detected, we can enroll it automatically.

**Review by aplanas — COMMENTED** [review]
not tested, but mostly LGTM, but I am a bit worried about the interactivity-only side

File: jeos-firstboot-enroll (hunk 15 above)
Comment (lnussel) [reply jeos-firstboot-enroll:57]: Huh? This is the firstboot *wizard*. The whole point of this thing is to be interactive. The autodeployment case is meant to be handled by combustion. When combustion is used, no firstboot wizard is started.

Having said that I've also implemented credential support for the wizard which can actually make it non-interactive: https://github.com/openSUSE/jeos-firstboot/pull/119
Once that's in we can implement credential support also for the enroll module.

File: disk-encryption-tool-dracut (hunk 12 above)
Comment (lnussel) [reply disk-encryption-tool-dracut:12]: no idea. according to my tests it's launched via the service file.

File: jeos-firstboot-enroll (hunk 15 above)
Comment (aplanas) [reply jeos-firstboot-enroll:57]: > Huh? This is the firstboot wizard. 

wizard is a too strong word for the jeos firstboot modules ...

> The autodeployment case is meant to be handled by combustion. When combustion is used, no firstboot wizard is started.

OK then.

**Review by aplanas — APPROVED** [review]

**Pull request merged**